File: TunnelProject/Opc/OpcItemDictionary.cs
using System.Linq;
using TunnelProject.Opc.Interfaces;

namespace TunnelProject.Opc
{
    internal class OpcItemDictionary : Dictionary<OpcItemKey, IOpcObject>
    {
        /// <summary>
        /// Gets the OPC object by PLC tag name.
        /// Throws if not found.
        /// </summary>
        public IOpcObject GetByPlcTagName(string plcTagName)
        {
            var key = this.Keys.FirstOrDefault(k => k.PlcTagName == plcTagName);
            if (key != null)
                return this[key];
            throw new ArgumentException($"PLC tag name '{plcTagName}' does not exist.");
        }

        /// <summary>
        /// Gets the OPC object by client handle.
        /// Throws if not found.
        /// </summary>
        public IOpcObject GetByClientHandle(int clientHandle)
        {
            var key = this.Keys.FirstOrDefault(k => k.ClientHandle == clientHandle);
            if (key != null)
                return this[key];
            throw new ArgumentException($"Client handle '{clientHandle}' does not exist.");
        }

        /// <summary>
        /// Tries to get the OPC object by PLC tag name.
        /// </summary>
        public bool TryGetByPlcTagName(string plcTagName, out IOpcObject value)
        {
            var key = this.Keys.FirstOrDefault(k => k.PlcTagName == plcTagName);
            if (key != null)
            {
                value = this[key];
                return true;
            }
            value = default;
            return false;
        }

        /// <summary>
        /// Tries to get the OPC object by client handle.
        /// </summary>
        public bool TryGetByClientHandle(int clientHandle, out IOpcObject value)
        {
            var key = this.Keys.FirstOrDefault(k => k.ClientHandle == clientHandle);
            if (key != null)
            {
                value = this[key];
                return true;
            }
            value = default;
            return false;
        }

        /// <summary>
        /// Checks if a PLC tag name exists.
        /// </summary>
        public bool ContainsPlcTagName(string plcTagName)
        {
            return this.Keys.Any(k => k.PlcTagName == plcTagName);
        }

        /// <summary>
        /// Checks if a client handle exists.
        /// </summary>
        public bool ContainsClientHandle(int clientHandle)
        {
            return this.Keys.Any(k => k.ClientHandle == clientHandle);
        }

        /// <summary>
        /// Gets the OpcItemKey by client handle.
        /// Throws if not found.
        /// </summary>
        public OpcItemKey GetKeyByClientHandle(int clientHandle)
        {
            var key = this.Keys.FirstOrDefault(k => k.ClientHandle == clientHandle);
            if (key != null)
                return key;
            throw new ArgumentException($"Client handle '{clientHandle}' is not defined.");
        }

        /// <summary>
        /// Gets the OpcItemKey by PLC tag name.
        /// Throws if not found.
        /// </summary>
        public OpcItemKey GetKeyByPlcTagName(string plcTagName)
        {
            var key = this.Keys.FirstOrDefault(k => k.PlcTagName == plcTagName);
            if (key != null)
                return key;
            throw new ArgumentException($"PLC tag name '{plcTagName}' is not defined.");
        }
    }
}

File: TSA_Rail_System_TunnelProject/Wdw_Sust.OpcDa.Core/OpcItemDictionary.vb
Imports Wdw_Sust.OpcDa.Core.Interfaces

Friend Class OpcItemDictionary
    Inherits Dictionary(Of OpcItemKey, IOpcObject)

    Public Shadows Function ContainsKey(key As Integer) As Boolean
        For Each i As OpcItemKey In Keys
            If i.ClientHandle = key Then
                Return True
            End If
        Next
        Return False
    End Function

    Public Shadows Function ContainsKey(key As String) As Boolean
        For Each s As OpcItemKey In Keys
            If s.PlcTagName = key Then
                Return True
            End If
        Next
        Return False
    End Function

    Public Shadows Function ContainsKey(key As OpcItemKey) As Boolean
        Return MyBase.ContainsKey(key)
    End Function

    Public Function GetKey(clientHandle As Integer) As OpcItemKey
        For Each tempKey As OpcItemKey In Keys
            If tempKey.ClientHandle = clientHandle Then
                Return tempKey
            End If
        Next
        Throw New ArgumentException("This client handle is not defined!!")
    End Function

    Public Function GetKey(plcTagName As String) As OpcItemKey
        For Each tempKey As OpcItemKey In Keys
            If tempKey.PlcTagName = plcTagName Then
                Return tempKey
            End If
        Next
        Throw New ArgumentException(String.Format("PLC tag name {0} is not defined!!", plcTagName))
    End Function

    ''' <summary>
    ''' Used to get the OPC object by using the PLC tag name of the object
    ''' </summary>
    ''' <param name="key">Name of the PLC tag name</param>
    ''' <returns>The IOPCObject from the list</returns>
    ''' <remarks></remarks>
    Public Shadows Function Item(key As String) As IOpcObject
        For Each s As OpcItemKey In Keys
            If s.PlcTagName = key Then
                Return MyBase.Item(s)
            End If
        Next
        Throw New ArgumentException(String.Format("Key {0} Does Not Exist", key))
    End Function

    ''' <summary>
    ''' Used to get the opc object by using the client handle of the opc object
    ''' </summary>
    ''' <param name="key">Client handle of the opc item</param>
    ''' <returns>The IOPCObject from the list</returns>
    ''' <remarks></remarks>
    Public Shadows Function Item(key As Integer) As IOpcObject
        For Each i As OpcItemKey In Keys
            If i.ClientHandle = key Then
                Return MyBase.Item(i)
            End If
        Next
        Throw New ArgumentException("Key Does Not Exist")
    End Function

    Public Shadows Sub Remove(key As OpcItemKey)
        Me(key) = Nothing
        MyBase.Remove(key)
    End Sub

End Class